samples/WinForms/Dx12Panel.cs

```diff
@@ -0,0 +1,868 @@
+using System.Runtime.InteropServices;
+using System.Windows.Forms;
+using TerraFX.Interop;
+using static TerraFX.Interop.D3D_FEATURE_LEVEL;
+using static TerraFX.Interop.D3D_PRIMITIVE_TOPOLOGY;
+using static TerraFX.Interop.D3D_ROOT_SIGNATURE_VERSION;
+using static TerraFX.Interop.D3D12;
+using static TerraFX.Interop.D3D12_COMMAND_LIST_TYPE;
+using static TerraFX.Interop.D3D12_DESCRIPTOR_HEAP_TYPE;
+using static TerraFX.Interop.D3D12_FENCE_FLAGS;
+using static TerraFX.Interop.D3D12_HEAP_FLAGS;
+using static TerraFX.Interop.D3D12_HEAP_TYPE;
+using static TerraFX.Interop.D3D12_INPUT_CLASSIFICATION;
+using static TerraFX.Interop.D3D12_PRIMITIVE_TOPOLOGY_TYPE;
+using static TerraFX.Interop.D3D12_RESOURCE_STATES;
+using static TerraFX.Interop.D3D12_ROOT_SIGNATURE_FLAGS;
+using static TerraFX.Interop.D3DCompiler;
+using static TerraFX.Interop.DXGI;
+using static TerraFX.Interop.DXGI_ADAPTER_FLAG;
+using static TerraFX.Interop.DXGI_FORMAT;
+using static TerraFX.Interop.DXGI_SWAP_EFFECT;
+using static TerraFX.Interop.Kernel32;
+using static TerraFX.Interop.Windows;
+using static TerraFX.Samples.DirectX.DXSampleHelper;
+
+
+namespace TerraFX.Samples.WinForms
+{
+    public unsafe class DX12Panel : Panel
+    {
+        #region local structs
+
+        private struct Vertex
+        {
+            public Vector3 Position;
+
+            public Vector4 Color;
+        }
+
+        #endregion local structs
+
+        #region data members
+
+        private float _aspectRatio;
+
+        // Adapter info
+        private readonly bool _useWarpDevice;
+
+        // Root assets path
+        private readonly string _assetsPath;
+
+        private const uint FrameCount = 2;
+
+        // Pipeline objects
+        private D3D12_VIEWPORT _viewport;
+        private RECT _scissorRect;
+        private IDXGISwapChain3* _swapChain;
+        private ID3D12Device* _device;
+        private readonly ID3D12Resource*[] _renderTargets;
+        private ID3D12CommandAllocator* _commandAllocator;
+        private ID3D12CommandQueue* _commandQueue;
+        private ID3D12RootSignature* _rootSignature;
+        private ID3D12DescriptorHeap* _rtvHeap;
+        private ID3D12PipelineState* _pipelineState;
+        private ID3D12GraphicsCommandList* _commandList;
+        private uint _rtvDescriptorSize;
+        private bool _isOnColorToggle;
+
+        // App resources.
+        private ID3D12Resource* _vertexBuffer;
+        private D3D12_VERTEX_BUFFER_VIEW _vertexBufferView;
+
+        // Synchronization objects.
+        private uint _frameIndex;
+        private HANDLE _fenceEvent;
+        private ID3D12Fence* _fence;
+        private ulong _fenceValue;
+
+        public float AspectRatio => _aspectRatio;
+
+        public bool UseWarpDevice => _useWarpDevice;
+
+        private Vector2 _mousePriorPt;
+
+        #endregion data members
+
+        #region constructor
+
+        public DX12Panel(int width, int height, bool useWarpDevice = false) : base()
```

Comment:
Note to self: There is a lot of duplication between this type and the underlying non-WinForms samples. Ideally we could have a shared abstraction that `DX12Panel` just wraps and we could support rendering any of the samples in WinForms with no changes required.

This would also simplify things when the sample is extended to WPF or other UI frameworks.

Reply:
Agreed.  I think it stems from the original C++ examples that have a lot of shared code, too. They seem to have been made via copy and modify.  But now that they are that way its probably best to have the C# version be as similar as possible. So within TerraFX.Interop.Windows.sln the duplication is fine.  But still in TerraFX.Interop.Windows.Samples.sln where there is no C++ reference reuse would be better.  Something to refactor when we have the second one?

Reply:
I was actually thinking of just tweaking the class that holds the sample slightly so it could be reused here as well. The general port would still match but each sample, with minimal modification, could be tweaked so it can be hosted anywhere.

I'll likely prototype something here myself.

Reply:
Sure, please tweak it.  
Also see https://github.com/terrafx/terrafx.interop.windows/issues/61 that addresses the other similar concern you had.  I suppose for now that means no further change on this PR due to this particular problem.